File: admin-invite-email-collision/python/check_invite_collision.py
"""Flag a Medusa v2 admin invite that will fail at accept time because the
target email already has a customer AuthIdentity. Both customer registration
and invite acceptance call the same auth provider register method, which is
keyed only on email with no actor_type awareness, so a customer identity on
that email guarantees a 401 Identity with email already exists when the
invite is accepted. This never mutates an existing identity. DRY_RUN=true
only reports the collision check. Safe to run again and again.

Guide: https://www.allanninal.dev/medusa/admin-invite-email-collision/
"""
import os
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("check_invite_collision")

BASE_URL = os.environ.get("MEDUSA_BACKEND_URL", "http://localhost:9000")
EMAIL = os.environ.get("MEDUSA_ADMIN_EMAIL", "admin@example.com")
PASSWORD = os.environ.get("MEDUSA_ADMIN_PASSWORD", "supersecret")
TARGET_EMAIL = os.environ.get("TARGET_EMAIL", "jane@example.com")
INVITE_ROLE = os.environ.get("INVITE_ROLE", "admin")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"


def get_token():
    r = requests.post(
        f"{BASE_URL}/auth/user/emailpass",
        json={"email": EMAIL, "password": PASSWORD},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["token"]


def find_customers(token, email):
    headers = {"Authorization": f"Bearer {token}"}
    r = requests.get(
        f"{BASE_URL}/admin/customers",
        params={"email": email, "fields": "id,email,has_account"},
        headers=headers,
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["customers"]


def find_admin_users(token, email):
    headers = {"Authorization": f"Bearer {token}"}
    r = requests.get(
        f"{BASE_URL}/admin/users",
        params={"email": email, "fields": "id,email"},
        headers=headers,
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["users"]


def list_pending_invites(token):
    headers = {"Authorization": f"Bearer {token}"}
    r = requests.get(
        f"{BASE_URL}/admin/invites",
        params={"fields": "id,email,accepted,expires_at"},
        headers=headers,
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["invites"]


def will_invite_collide(target_email, customers, admin_users, pending_invites):
    """Pure: no I/O. Returns {"safe": bool, "reason": str}.

    reason is one of "customer_account_exists", "admin_user_exists",
    "invite_pending", or "ok".
    """
    email = target_email.strip().lower()

    for customer in customers:
        if customer.get("email", "").strip().lower() == email and customer.get("has_account") is True:
            return {"safe": False, "reason": "customer_account_exists"}

    for admin_user in admin_users:
        if admin_user.get("email", "").strip().lower() == email:
            return {"safe": False, "reason": "admin_user_exists"}

    for invite in pending_invites:
        if invite.get("email", "").strip().lower() == email and invite.get("accepted") is not True:
            return {"safe": False, "reason": "invite_pending"}

    return {"safe": True, "reason": "ok"}


def create_invite(token, email, role):
    headers = {"Authorization": f"Bearer {token}"}
    r = requests.post(
        f"{BASE_URL}/admin/invites",
        json={"email": email, "role": role},
        headers=headers,
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["invite"]


def run():
    token = get_token()
    customers = find_customers(token, TARGET_EMAIL)
    admin_users = find_admin_users(token, TARGET_EMAIL)
    pending_invites = list_pending_invites(token)

    decision = will_invite_collide(TARGET_EMAIL, customers, admin_users, pending_invites)

    if not decision["safe"]:
        log.warning(
            "Blocked: invite to %s would collide (%s). Invite a different email instead.",
            TARGET_EMAIL, decision["reason"],
        )
        return

    log.info("Email %s is clear. %s", TARGET_EMAIL, "would create invite" if DRY_RUN else "creating invite")
    if not DRY_RUN:
        invite = create_invite(token, TARGET_EMAIL, INVITE_ROLE)
        log.info("Invite created: %s", invite["id"])


if __name__ == "__main__":
    run()

File: admin-invite-email-collision/node/check-invite-collision.js
/**
 * Flag a Medusa v2 admin invite that will fail at accept time because the
 * target email already has a customer AuthIdentity. Both customer registration
 * and invite acceptance call the same auth provider register method, which is
 * keyed only on email with no actor_type awareness, so a customer identity on
 * that email guarantees a 401 Identity with email already exists when the
 * invite is accepted. This never mutates an existing identity. DRY_RUN=true
 * only reports the collision check. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/medusa/admin-invite-email-collision/
 */
import { pathToFileURL } from "node:url";

const BASE_URL = process.env.MEDUSA_BACKEND_URL || "http://localhost:9000";
const EMAIL = process.env.MEDUSA_ADMIN_EMAIL || "admin@example.com";
const PASSWORD = process.env.MEDUSA_ADMIN_PASSWORD || "supersecret";
const TARGET_EMAIL = process.env.TARGET_EMAIL || "jane@example.com";
const INVITE_ROLE = process.env.INVITE_ROLE || "admin";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

export function willInviteCollide(targetEmail, customers, adminUsers, pendingInvites) {
  // Pure: no I/O. Returns { safe, reason }.
  // reason is one of "customer_account_exists", "admin_user_exists",
  // "invite_pending", or "ok".
  const email = targetEmail.trim().toLowerCase();

  const customerHit = customers.find(
    (c) => (c.email || "").trim().toLowerCase() === email && c.has_account === true
  );
  if (customerHit) return { safe: false, reason: "customer_account_exists" };

  const adminHit = adminUsers.find((u) => (u.email || "").trim().toLowerCase() === email);
  if (adminHit) return { safe: false, reason: "admin_user_exists" };

  const inviteHit = pendingInvites.find(
    (i) => (i.email || "").trim().toLowerCase() === email && i.accepted !== true
  );
  if (inviteHit) return { safe: false, reason: "invite_pending" };

  return { safe: true, reason: "ok" };
}

async function getToken() {
  const res = await fetch(`${BASE_URL}/auth/user/emailpass`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ email: EMAIL, password: PASSWORD }),
  });
  if (!res.ok) throw new Error(`Medusa auth ${res.status}`);
  const body = await res.json();
  return body.token;
}

async function apiGet(token, path, params) {
  const url = new URL(`${BASE_URL}${path}`);
  for (const [key, value] of Object.entries(params)) url.searchParams.set(key, value);
  const res = await fetch(url, { headers: { Authorization: `Bearer ${token}` } });
  if (!res.ok) throw new Error(`Medusa ${path} ${res.status}`);
  return res.json();
}

async function findCustomers(token, email) {
  const body = await apiGet(token, "/admin/customers", { email, fields: "id,email,has_account" });
  return body.customers;
}

async function findAdminUsers(token, email) {
  const body = await apiGet(token, "/admin/users", { email, fields: "id,email" });
  return body.users;
}

async function listPendingInvites(token) {
  const body = await apiGet(token, "/admin/invites", { fields: "id,email,accepted,expires_at" });
  return body.invites;
}

async function createInvite(token, email, role) {
  const res = await fetch(`${BASE_URL}/admin/invites`, {
    method: "POST",
    headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json" },
    body: JSON.stringify({ email, role }),
  });
  if (!res.ok) throw new Error(`Medusa ${res.status}`);
  const body = await res.json();
  return body.invite;
}

export async function run() {
  const token = await getToken();
  const [customers, adminUsers, pendingInvites] = await Promise.all([
    findCustomers(token, TARGET_EMAIL),
    findAdminUsers(token, TARGET_EMAIL),
    listPendingInvites(token),
  ]);

  const decision = willInviteCollide(TARGET_EMAIL, customers, adminUsers, pendingInvites);

  if (!decision.safe) {
    console.warn(
      `Blocked: invite to ${TARGET_EMAIL} would collide (${decision.reason}). Invite a different email instead.`
    );
    return;
  }

  console.log(`Email ${TARGET_EMAIL} is clear. ${DRY_RUN ? "would create invite" : "creating invite"}`);
  if (!DRY_RUN) {
    const invite = await createInvite(token, TARGET_EMAIL, INVITE_ROLE);
    console.log(`Invite created: ${invite.id}`);
  }
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
